File: Programmers/[72416] 매출 하락 최소화/SUbbb/72416.java
import java.util.*;

class Solution {
    // 직원별 워크숍에 참여했을 때의 매출과 참여하지 않았을 때의 매출, 0: 참여 X, 1: 참여 O
    static int[][] costs;
    static ArrayList<ArrayList<Integer>> adjList = new ArrayList<>();
    static int[] sales;
    
    public int solution(int[] sales, int[][] links) {
        this.sales = sales;
        init();
        
        makeAdjList(links);
        
        traversal(0);
        
        return Math.min(costs[0][0], costs[0][1]);
    }
    
    private static void init() {
        costs = new int[sales.length][2];
        
        for (int i = 0; i < sales.length; i++) 
            adjList.add(new ArrayList<>());
    }
    
    private static void makeAdjList(int[][] links) {
        for (int[] link : links)
            adjList.get(link[0] - 1).add(link[1] - 1);
    }
    
    private static void traversal(int node) {
        // 해당 직원이 워크숍에 참여하지 않는 경우
        costs[node][0] = 0;
        // 해당 직원이 워크숍에 참여한 경우 -> 해당 직원 매출을 저장
        costs[node][1] = sales[node];
        
        // 리프 노드라면, 상위로 최소 비용을 전달해야 함
        if (adjList.get(node).isEmpty()) return;
        
        int extraCost = 10001;
        for (int child : adjList.get(node)) {
            traversal(child);
            
            // 최소 비용을 부모 노드(팀장)에 누적
            if (costs[child][0] < costs[child][1]) {
                costs[node][0] += costs[child][0];
                costs[node][1] += costs[child][0];
                // 팀장과 팀원 모두 참여하지 않으면 안되기에 필요한 추가 비용 계산
                extraCost = Math.min(extraCost, costs[child][1] - costs[child][0]);
            } else {
                costs[node][0] += costs[child][1];
                costs[node][1] += costs[child][1];    
                // 둘 중 한 명이라도 참여했다면 추가 비용은 필요 없음
                extraCost = 0;
            }
        }
        
        costs[node][0] += extraCost;
    }
}

File: Programmers/[72416] 매출 하락 최소화/Johoseong/72416.py
def dfs(graph, sales, DP, node):
    if not graph[node]: # 리프노드(=그냥 팀원)이면 선택 시 비용 자기자신임. 바로 return
        DP[node][0] = sales[node]
        return

    sum_child = 0
    for child in graph[node]: # 한 팀의 최소 비용 구함 (그 밑에 팀 있다면 그거까지)
        dfs(graph, sales, DP, child)
        sum_child += min(DP[child][0], DP[child][1])
    DP[node][0] = sum_child + sales[node]

    attend = False
    for child in graph[node]:
        if DP[child][1] > DP[child][0]: # 이미 팀원이 다른팀의 팀장으로서 참석하는 경우!!
            attend = True
    if attend: 
        DP[node][1] = sum_child
    else: 
        DP[node][1] = sum_child + min((DP[k][0] - DP[k][1] for k in graph[node]), default=0)

def solution(sales, links):
    N = len(sales)
    graph = { (i + 1):[] for i in range(N)} # 루트는 항상 1
    sales = [0] + sales
    for a, b in links:
        graph[a].append(b)
    DP = [0] + [[0, 0] for _ in range(N)]

    dfs(graph, sales, DP, 1)

    return min(DP[1][0], DP[1][1])

print(solution([14, 17, 15, 18, 19, 14, 13, 16, 28, 17], [[10, 8], [1, 9], [9, 7], [5, 4], [1, 5], [5, 10], [10, 6], [1, 3], [10, 2]]))